Lora syncer docs (#320)
kubernetes-sigs/gateway-api-inference-extension: site-src/guides/dynamic-lora.md

# Getting started with Gateway API Inference Extension with Dynamic lora updates on vllm

The goal of this guide is to get a single InferencePool running with VLLM and demonstrate use of dynamic lora updating ! 

### Requirements
 - Envoy Gateway [v1.2.1](https://gateway.envoyproxy.io/docs/install/install-yaml/#install-with-yaml) or higher
 - A cluster with:
   - Support for Services of type `LoadBalancer`. (This can be validated by ensuring your Envoy Gateway is up and running). For example, with Kind,
     you can follow [these steps](https://kind.sigs.k8s.io/docs/user/loadbalancer).
   - 3 GPUs to run the sample model server. Adjust the number of replicas in `./manifests/vllm/deployment.yaml` as needed.

### Steps

1. **Deploy Sample VLLM Model Server with dynamic lora update enabled and dynamic lora syncer sidecar **

>     [Deploy sample vllm deployment with Dynamic lora adapter enabled and Lora syncer sidecar and configmap](https://github.com/kubernetes-sigs/gateway-api-inference-extension/blob/main/pkg/manifests/vllm/dynamic-lora-sidecar/deployment.yaml)

Reviewer: can you adjust this one please

Rest of the steps are same as [general setup](https://github.com/kubernetes-sigs/gateway-api-inference-extension/blob/main/site-src/guides/index.md)

### Safely rollout v2 adapter
    

> 1. Update lora configmap

Reviewer: ditto

``` yaml

        apiVersion: v1
        kind: ConfigMap
        metadata:
        name: dynamic-lora-config
        data:
        configmap.yaml: |
            vLLMLoRAConfig:
            ensureExist:   
                models:
                - id: tweet-summary-v1
                    source: tweet-summary-1=/adapters/vineetsharma/qlora-adapter-Llama-2-7b-hf-TweetSumm_1
                - id: tweet-summary-v2

>                     source: tweet-summary-2=/adapters/vineetsharma/qlora-adapter-Llama-2-7b-hf-TweetSumm_2

Reviewer: needs to be tweet-summary-0, tweet-summary-1 (existing) and a new tweet-summary-2

Reviewer: also, to update the configmap, we can instruct the user to:

```
kubectl edit dynamic-lora-config
```

    ```

> 2. Configure a canary rollout with traffic split using LLMService. In this example, 10% of traffic to the chatbot model will be sent to v2.

Reviewer: ```suggestion
2. Configure a canary rollout with traffic split using InferenceModel. In this example, 10% of traffic to the chatbot model will be sent to `tweet-summary-3`.
```

``` yaml
model:
    name: chatbot
    targetModels:
    targetModelName: chatbot-v1
            weight: 90
    targetModelName: chatbot-v2
            weight: 10

> ```

Reviewer: This needs to match the update we asked the user to do on the configmap, so here we should do `tweet-summary-0`, `tweet-summary-1` (exiting) and we add tweet-summary-2.

We can tell the user to do:
`kubectl edit inferencemodel-sample` 

to edit the existing InferenceModel object.

            
3. Finish rollout by setting the traffic to the new version 100%.
```yaml
model:
    name: chatbot
    targetModels:
    targetModelName: chatbot-v2
            weight: 100
```
    
4. Remove v1 from dynamic lora configmap.
```yaml
    apiVersion: v1
    kind: ConfigMap
    metadata:
    name: dynamic-lora-config
    data:
    configmap.yaml: |
            vLLMLoRAConfig:
            ensureExist:
            models:
            - id: chatbot-v2

>                 source: gs://[TEAM-A-MODELS-BUCKET]/chatbot-v2

Reviewer: need to update those, lets only work with hugging face adapters, otherwise it will be hard to run through the guide

   - Support for Services of type `LoadBalancer`. (This can be validated by ensuring your Envoy Gateway is up and running). For example, with Kind,
     you can follow [these steps](https://kind.sigs.k8s.io/docs/user/loadbalancer).
   - 3 GPUs to run the sample model server. Adjust the number of replicas in `./manifests/vllm/deployment.yaml` as needed.

### Steps

1. **Deploy Sample VLLM Model Server with dynamic lora update enabled and dynamic lora syncer sidecar **
    [Deploy sample vllm deployment with Dynamic lora adapter enabled and Lora syncer sidecar and configmap](https://github.com/kubernetes-sigs/gateway-api-inference-extension/blob/main/pkg/manifests/vllm/dynamic-lora-sidecar/deployment.yaml)

Rest of the steps are same as [general setup](https://github.com/kubernetes-sigs/gateway-api-inference-extension/blob/main/site-src/guides/index.md)

### Safely rollout v2 adapter
    
1. Update lora configmap

``` yaml

        apiVersion: v1
        kind: ConfigMap
        metadata:
        name: dynamic-lora-config
        data:
        configmap.yaml: |
             vLLMLoRAConfig:
                name: sql-loras-llama
                port: 8000
                ensureExist:

>                     models:

Reviewer: need to add tweet-summary-2 to this

Author: Done

                      id: tweet-summary-0
                      source: vineetsharma/qlora-adapter-Llama-2-7b-hf-TweetSumm
                    - base-model: meta-llama/Llama-2-7b-hf
                      id: tweet-summary-1
                      source: vineetsharma/qlora-adapter-Llama-2-7b-hf-TweetSumm
                    - base-model: meta-llama/Llama-2-7b-hf
                      id: tweet-summary-2

>                       source: vineetsharma/qlora-adapter-Llama-2-7b-hf-TweetSumm

Reviewer: ```suggestion
                      source: vineetsharma/qlora-adapter-Llama-2-7b-hf-TweetSumm
```                 
```

>                       source: vineetsharma/qlora-adapter-Llama-2-7b-hf-TweetSumm

Reviewer: need to close this with ```, I couldn't make a suggestion to fix that.

2. Configure a canary rollout with traffic split using LLMService. In this example, 40% of traffic for tweet-summary model will be sent to the ***tweet-summary-2*** adapter .

``` yaml
model:
    name: tweet-summary
    targetModels:
    targetModelName: tweet-summary-0
            weight: 10
    targetModelName: tweet-summary-1
            weight: 40
    targetModelName: tweet-summary-2

>             weight: 40

Reviewer: they don't add up to a 100, they don't need to, but it is easier to reason about if they do.

Author: Fixed

Rest of the steps are same as [general setup](https://github.com/kubernetes-sigs/gateway-api-inference-extension/blob/main/site-src/guides/index.md)

### Safely rollout v2 adapter
    
1. Update the LoRA syncer ConfigMap to make the new adapter version available on the model servers.

```yaml
        apiVersion: v1
        kind: ConfigMap
        metadata:
        name: dynamic-lora-config
        data:
        configmap.yaml: |
             vLLMLoRAConfig:
                name: sql-loras-llama
                port: 8000
                ensureExist:
                    models:
                    - base-model: meta-llama/Llama-2-7b-hf
                      id: tweet-summary-0
                      source: vineetsharma/qlora-adapter-Llama-2-7b-hf-TweetSumm
                    - base-model: meta-llama/Llama-2-7b-hf
                      id: tweet-summary-1
                      source: vineetsharma/qlora-adapter-Llama-2-7b-hf-TweetSumm
                    - base-model: meta-llama/Llama-2-7b-hf
                      id: tweet-summary-2
                      source: vineetsharma/qlora-adapter-Llama-2-7b-hf-TweetSumm

> 

Reviewer: ```suggestion
```

```

The goal of this guide is to get a single InferencePool running with vLLM and demonstrate use of dynamic lora updating! 

### Requirements
 - Envoy Gateway [v1.2.1](https://gateway.envoyproxy.io/docs/install/install-yaml/#install-with-yaml) or higher
 - A cluster with:
   - Support for Services of type `LoadBalancer`. (This can be validated by ensuring your Envoy Gateway is up and running). For example, with Kind,
     you can follow [these steps](https://kind.sigs.k8s.io/docs/user/loadbalancer).
   - 3 GPUs to run the sample model server. Adjust the number of replicas in `./manifests/vllm/deployment.yaml` as needed.

### Steps

1. **Deploy Sample VLLM Model Server with dynamic lora update enabled and dynamic lora syncer sidecar **
    [Deploy sample vllm deployment with Dynamic lora adapter enabled and Lora syncer sidecar and configmap](https://github.com/kubernetes-sigs/gateway-api-inference-extension/blob/main/pkg/manifests/vllm/dynamic-lora-sidecar/deployment.yaml)

Rest of the steps are same as [general setup](https://github.com/kubernetes-sigs/gateway-api-inference-extension/blob/main/site-src/guides/index.md)

### Safely rollout v2 adapter
    
1. Update lora configmap

> ``` yaml

Reviewer: ```suggestion
```yaml
```

        apiVersion: v1
        kind: ConfigMap
        metadata:
        name: dynamic-lora-config
        data:
        configmap.yaml: |
             vLLMLoRAConfig:
                name: sql-loras-llama
                port: 8000
                ensureExist:
                    models:
                    - base-model: meta-llama/Llama-2-7b-hf
                      id: tweet-summary-0
                      source: vineetsharma/qlora-adapter-Llama-2-7b-hf-TweetSumm
                    - base-model: meta-llama/Llama-2-7b-hf
                      id: tweet-summary-1
                      source: vineetsharma/qlora-adapter-Llama-2-7b-hf-TweetSumm
                    - base-model: meta-llama/Llama-2-7b-hf
                      id: tweet-summary-2
                      source: vineetsharma/qlora-adapter-Llama-2-7b-hf-TweetSumm

>     ```

Reviewer: ```suggestion
```
```

2. Configure a canary rollout with traffic split using LLMService. In this example, 40% of traffic for tweet-summary model will be sent to the ***tweet-summary-2*** adapter .

> ``` yaml

Reviewer: ```suggestion
```yaml
```
